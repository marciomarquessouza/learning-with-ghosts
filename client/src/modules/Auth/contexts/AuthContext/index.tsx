import React, { createContext, useEffect, useState, useCallback } from 'react'

import { User } from 'types'
import { useDocument } from 'common/hooks/useDocument'
import { COLLECTIONS } from 'const'
import { useAuth } from 'modules/Auth/hooks/useAuth'

export interface AuthContextType {
    user?: User | null
    loading: boolean
    error?: Error | null
    setChapter(chapter: number): Promise<null | undefined>
}

interface AuthProviderProps {
    children: React.ReactNode
}

export const AuthContext = createContext<AuthContextType>({
    user: null,
    loading: false,
    error: null,
    setChapter: async (chapter: number) => null,
})

function AuthProvider({ children }: AuthProviderProps) {
    const { user: authUser, loading: authUserLoading } = useAuth()
    const [user, setUser] = useState<User | null>()
    const [loading, setLoading] = useState(true)
    const [error, setError] = useState(null)
    const { getDocument, setDocument } = useDocument()

    const handleSetChapter = useCallback(
        async (chapter: number) => {
            if (!user) return null
            setLoading(true)
            setDocument<User>(COLLECTIONS.USERS, user.uid, { chapter })
                .then((updatedUser) => {
                    setUser(updatedUser)
                })
                .catch((reason) => {
                    setError(reason)
                })
                .finally(() => {
                    setLoading(false)
                })
        },
        [setDocument, user]
    )

    useEffect(() => {
        if (authUserLoading) return
        if (authUser) {
            getDocument<User>(COLLECTIONS.USERS, authUser.uid)
                .then((userData) => {
                    setUser(userData)
                    setLoading(false)
                })
                .catch((reason) => {
                    setError(reason)
                })
        } else {
            setUser(null)
            setLoading(false)
        }
    }, [getDocument, authUser, authUserLoading])

    return (
        <AuthContext.Provider
            value={{
                user,
                loading,
                error,
                setChapter: handleSetChapter,
            }}
        >
            {children}
        </AuthContext.Provider>
    )
}

export default AuthProvider
