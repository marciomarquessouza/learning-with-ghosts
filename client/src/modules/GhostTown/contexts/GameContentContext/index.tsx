import React, { createContext, useCallback } from 'react'
import { useQuery } from 'react-query'

import { Chapter } from 'types'
import { fetchGameContent } from './fetchGameContent'
import { useGameProgress } from 'modules/GhostTown/hooks/useGameProgress'

export interface GameContentContextType {
    chapter?: Chapter | null
    status: 'loading' | 'error' | 'idle' | 'success'
    error: boolean
}

interface GameContentProviderProps {
    children: React.ReactNode
}

export const GameContentContext = createContext<GameContentContextType>({
    chapter: null,
    status: 'idle',
    error: false,
})

function GameContentProvider({ children }: GameContentProviderProps) {
    const { gameProgress, isReady: isGameProgressReady } = useGameProgress()
    const { data, status, isError } = useQuery({
        queryKey: ['interactions', gameProgress.chapter],
        queryFn: () => fetchGameContent(gameProgress.chapter),
        enabled: isGameProgressReady,
    })

    return (
        <GameContentContext.Provider
            value={{
                chapter: data,
                error: isError,
                status: status,
            }}
        >
            {children}
        </GameContentContext.Provider>
    )
}

export default GameContentProvider
