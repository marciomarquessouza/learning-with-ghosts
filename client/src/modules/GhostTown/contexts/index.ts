export * from './GhostTownGUIContext'
export * from './InteractionContext'
