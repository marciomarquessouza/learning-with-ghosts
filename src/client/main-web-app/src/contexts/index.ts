export { default as AlertContext } from './AlertContext'
