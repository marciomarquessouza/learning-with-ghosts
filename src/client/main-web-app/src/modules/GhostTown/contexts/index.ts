export * from './GhostTownGUIContext/Context'
export * from './GhostTownGUIContext/Provider'
