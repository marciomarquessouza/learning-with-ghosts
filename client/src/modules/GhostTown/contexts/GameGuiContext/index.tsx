import React, { createContext, useCallback, useReducer, useMemo, useEffect } from 'react'
import { useRouter } from 'next/router'

import { CHARACTER } from 'modules/GhostTown/const'
import { gameGuiReducer } from './reducers/gameGuiReducer'
import { gameGuiInitialState } from './reducers/gameGuiInitialState'
import { ACTIONS, MENUS, MenusProps } from 'types/GameGui'
import { menuSubject } from 'modules/GhostTown/observers'
import getActionByMenu from './utils/getActionByMenu'

import MenusWrapper from 'modules/GhostTown/components/MenusWrapper'
import { GAME_KEYS } from 'modules/GhostTown/player/controls'

export interface GameGuiProviderProps {
    children: React.ReactNode
}

export interface GameGuiContextType {
    openMenu: (menuProps: MenusProps) => void
    closeMenu: (menu: MENUS) => void
    callChallenge: (character?: CHARACTER) => void
    onKeyDown: (gameKeyInput: GAME_KEYS) => void
}

export const GameGuiContext = createContext({
    openMenu: (menuProps: MenusProps) => {},
    closeMenu: (menu: MENUS) => {},
    callChallenge: (character?: CHARACTER) => {},
    onKeyDown: (gameKeyInput: GAME_KEYS) => {},
})

function GameGuiProvider({ children }: GameGuiProviderProps) {
    const router = useRouter()
    const [state, dispatch] = useReducer(gameGuiReducer, gameGuiInitialState)

    const handleOpenMenu = (menuProps: MenusProps) => {
        const action = getActionByMenu(menuProps)
        dispatch(action)
    }

    const handleCloseMenu = (menu: MENUS) => {
        dispatch({ type: ACTIONS.CLOSE_MENU, menu })
    }

    const handleCallChallenge = (character?: CHARACTER) => {
        router.push('/lighthouse')
    }

    const handleOnKeyDown = (gameKeyInput: GAME_KEYS) => {
        dispatch({ type: ACTIONS.UPDATE_KEYS_INPUTS, value: gameKeyInput })
    }

    useEffect(() => {
        menuSubject.notifyDialogMenu(state.dialogMenu, state.isDialogMenuOpen)
        menuSubject.notifyInfoMenu(state.infoMenu, state.isInfoMenuOpen)
    }, [state.dialogMenu, state.infoMenu, state.isDialogMenuOpen, state.isInfoMenuOpen])

    return (
        <GameGuiContext.Provider
            value={{
                openMenu: handleOpenMenu,
                closeMenu: handleCloseMenu,
                callChallenge: handleCallChallenge,
                onKeyDown: handleOnKeyDown,
            }}
        >
            <MenusWrapper
                guiState={state}
                onOpenMenu={handleOpenMenu}
                onCloseMenu={handleCloseMenu}
            />
            {children}
        </GameGuiContext.Provider>
    )
}

export default GameGuiProvider
