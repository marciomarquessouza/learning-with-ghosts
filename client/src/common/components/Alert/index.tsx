import React, { Fragment } from 'react'
import { Transition } from '@headlessui/react'

import { ALERTS_TYPE_ENUM } from '../../contexts/AlertContext'

export interface AlertProps {
    title: string
    message: string
    type: ALERTS_TYPE_ENUM
    open: boolean
    onClose: () => void
}

function Alert({ title, message, type, open, onClose }: AlertProps) {
    const handleClose = (e: React.MouseEvent<HTMLElement>) => {
        e.preventDefault()
        onClose()
    }

    return (
        <div
            aria-live="assertive"
            className="pointer-events-none fixed inset-0 flex items-end px-4 py-6 sm:items-start sm:p-6"
        >
            <div className="flex w-full flex-col items-center space-y-4 sm:items-end">
                <Transition
                    show={open}
                    as={Fragment}
                    enter="transform ease-out duration-300 transition"
                    enterFrom="translate-y-2 opacity-0 sm:translate-y-0 sm:translate-x-2"
                    enterTo="translate-y-0 opacity-100 sm:translate-x-0"
                    leave="transition ease-in duration-100"
                    leaveFrom="opacity-100"
                    leaveTo="opacity-0"
                >
                    <div className="pointer-events-auto w-full max-w-sm overflow-hidden rounded-lg bg-white shadow-lg ring-1 ring-black ring-opacity-5">
                        <div className="p-4">
                            <div className="flex items-start">
                                <div className="flex-shrink-0">
                                    {type === ALERTS_TYPE_ENUM.SUCCESS ? (
                                        <svg
                                            width="15"
                                            height="15"
                                            viewBox="0 0 15 15"
                                            fill="none"
                                            xmlns="http://www.w3.org/2000/svg"
                                        >
                                            <circle cx="7.5" cy="7.5" r="7.5" fill="#34A26E" />
                                            <path
                                                d="M5.49686 9.92188L2.89061 7.31563L2.00311 8.19688L5.49686 11.6906L12.9969 4.19063L12.1156 3.30938L5.49686 9.92188Z"
                                                fill="white"
                                            />
                                        </svg>
                                    ) : (
                                        <svg
                                            xmlns="http://www.w3.org/2000/svg"
                                            width="20"
                                            height="20"
                                            viewBox="0 0 20 20"
                                        >
                                            <title>error</title>
                                            <path
                                                fill="#d33"
                                                d="M13.728 1H6.272L1 6.272v7.456L6.272 19h7.456L19 13.728V6.272zM11 15H9v-2h2zm0-4H9V5h2z"
                                            />
                                        </svg>
                                    )}
                                </div>
                                <div className="ml-3 w-0 flex-1 pt-0.5">
                                    <p className="text-sm font-semibold text-primary-light">
                                        {title}
                                    </p>
                                    <p className="mt-1 text-sm text-gray-500">{message}</p>
                                </div>
                                <div className="ml-4 flex flex-shrink-0">
                                    <button
                                        type="button"
                                        className="inline-flex rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
                                        onClick={handleClose}
                                    >
                                        <span className="sr-only">Close</span>
                                        <svg
                                            xmlns="http://www.w3.org/2000/svg"
                                            width="24"
                                            height="24"
                                            viewBox="0 0 24 24"
                                        >
                                            <g id="close">
                                                <path
                                                    id="x"
                                                    d="M18.717 6.697l-1.414-1.414-5.303 5.303-5.303-5.303-1.414 1.414 5.303 5.303-5.303 5.303 1.414 1.414 5.303-5.303 5.303 5.303 1.414-1.414-5.303-5.303z"
                                                />
                                            </g>
                                        </svg>
                                    </button>
                                </div>
                            </div>
                        </div>
                    </div>
                </Transition>
            </div>
        </div>
    )
}

export default Alert
