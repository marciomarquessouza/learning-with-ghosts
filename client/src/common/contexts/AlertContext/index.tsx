import React, { useState, createContext } from 'react'

import { MESSAGE_TIMEOUT } from '../../../const'

import Alert from '../../components/Alert'

export enum ALERTS_TYPE_ENUM {
    ERROR = 'error',
    SUCCESS = 'success',
}

export const AlertContext = createContext({
    openAlert: (props: { title: string; message: string; type: ALERTS_TYPE_ENUM }) => {},
    closeAlert: () => {},
})

export interface AlertProviderProps {
    children: React.ReactNode
}

function AlertProvider({ children }: AlertProviderProps) {
    const [open, setOpen] = useState(false)
    const [title, setTitle] = useState('')
    const [message, setMessage] = useState('')
    const [type, setType] = useState<ALERTS_TYPE_ENUM>(ALERTS_TYPE_ENUM.SUCCESS)

    const handleOpen = (props: { title: string; message: string; type: ALERTS_TYPE_ENUM }) => {
        setTitle(props.title)
        setMessage(props.message)
        setType(props.type)
        setOpen(true)
        setTimeout(() => {
            setOpen(false)
        }, MESSAGE_TIMEOUT)
    }

    const handleClose = () => {
        setOpen(false)
    }

    return (
        <AlertContext.Provider value={{ openAlert: handleOpen, closeAlert: handleClose }}>
            <Alert title={title} message={message} open={open} type={type} onClose={handleClose} />
            {children}
        </AlertContext.Provider>
    )
}

export default AlertProvider
