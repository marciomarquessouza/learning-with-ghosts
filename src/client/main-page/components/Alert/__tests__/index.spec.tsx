import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import Alert from '../'
import { ALERTS_TYPE_ENUM } from '../../../contexts/AlertContext'

const defaultMockedProps = {
    title: '_MOCK_TITLE_',
    message: '_MOCK_MESSAGE_',
    type: ALERTS_TYPE_ENUM.SUCCESS,
    open: true,
    onClose: jest.fn(),
}

describe('components >> Alert', () => {
    it('should render all types of Alerts properly', () => {
        const mockedSuccessProps = { ...defaultMockedProps, type: ALERTS_TYPE_ENUM.SUCCESS }
        const mockedErrorProps = { ...defaultMockedProps, type: ALERTS_TYPE_ENUM.ERROR }

        const { container: containerSuccess } = render(<Alert {...mockedSuccessProps} />)
        const { container: containerError } = render(<Alert {...mockedErrorProps} />)

        expect(containerSuccess).toMatchSnapshot('Alert-success')
        expect(containerError).toMatchSnapshot('Alert-error')
    })

    it('should fire the "onClose" function when the close button is pressed', () => {
        const spyOnClose = jest.fn()
        const mockedProps = { ...defaultMockedProps, onClose: spyOnClose }

        const { getByRole } = render(<Alert {...mockedProps} />)

        fireEvent.click(getByRole('button'))

        expect(spyOnClose).toHaveBeenCalledTimes(1)
    })
})
