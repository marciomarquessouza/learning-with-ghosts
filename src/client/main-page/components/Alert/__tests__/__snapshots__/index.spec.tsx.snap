// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components >> Alert should render all types of Alerts properly: Alert-error 1`] = `
<div>
  <div
    aria-live="assertive"
    class="pointer-events-none fixed inset-0 flex items-end px-4 py-6 sm:items-start sm:p-6"
  >
    <div
      class="flex w-full flex-col items-center space-y-4 sm:items-end"
    >
      <div
        class="pointer-events-auto w-full max-w-sm overflow-hidden rounded-lg bg-white shadow-lg ring-1 ring-black ring-opacity-5"
      >
        <div
          class="p-4"
        >
          <div
            class="flex items-start"
          >
            <div
              class="flex-shrink-0"
            >
              <svg
                height="20"
                viewBox="0 0 20 20"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <title>
                  error
                </title>
                <path
                  d="M13.728 1H6.272L1 6.272v7.456L6.272 19h7.456L19 13.728V6.272zM11 15H9v-2h2zm0-4H9V5h2z"
                  fill="#d33"
                />
              </svg>
            </div>
            <div
              class="ml-3 w-0 flex-1 pt-0.5"
            >
              <p
                class="text-sm font-semibold text-primary-light"
              >
                _MOCK_TITLE_
              </p>
              <p
                class="mt-1 text-sm text-gray-500"
              >
                _MOCK_MESSAGE_
              </p>
            </div>
            <div
              class="ml-4 flex flex-shrink-0"
            >
              <button
                class="inline-flex rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  Close
                </span>
                <svg
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    id="close"
                  >
                    <path
                      d="M18.717 6.697l-1.414-1.414-5.303 5.303-5.303-5.303-1.414 1.414 5.303 5.303-5.303 5.303 1.414 1.414 5.303-5.303 5.303 5.303 1.414-1.414-5.303-5.303z"
                      id="x"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`components >> Alert should render all types of Alerts properly: Alert-success 1`] = `
<div>
  <div
    aria-live="assertive"
    class="pointer-events-none fixed inset-0 flex items-end px-4 py-6 sm:items-start sm:p-6"
  >
    <div
      class="flex w-full flex-col items-center space-y-4 sm:items-end"
    >
      <div
        class="pointer-events-auto w-full max-w-sm overflow-hidden rounded-lg bg-white shadow-lg ring-1 ring-black ring-opacity-5"
      >
        <div
          class="p-4"
        >
          <div
            class="flex items-start"
          >
            <div
              class="flex-shrink-0"
            >
              <svg
                fill="none"
                height="15"
                viewBox="0 0 15 15"
                width="15"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="7.5"
                  cy="7.5"
                  fill="#34A26E"
                  r="7.5"
                />
                <path
                  d="M5.49686 9.92188L2.89061 7.31563L2.00311 8.19688L5.49686 11.6906L12.9969 4.19063L12.1156 3.30938L5.49686 9.92188Z"
                  fill="white"
                />
              </svg>
            </div>
            <div
              class="ml-3 w-0 flex-1 pt-0.5"
            >
              <p
                class="text-sm font-semibold text-primary-light"
              >
                _MOCK_TITLE_
              </p>
              <p
                class="mt-1 text-sm text-gray-500"
              >
                _MOCK_MESSAGE_
              </p>
            </div>
            <div
              class="ml-4 flex flex-shrink-0"
            >
              <button
                class="inline-flex rounded-md bg-white text-gray-400 hover:text-gray-500 focus:outline-none focus:ring-2 focus:ring-indigo-500 focus:ring-offset-2"
                type="button"
              >
                <span
                  class="sr-only"
                >
                  Close
                </span>
                <svg
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g
                    id="close"
                  >
                    <path
                      d="M18.717 6.697l-1.414-1.414-5.303 5.303-5.303-5.303-1.414 1.414 5.303 5.303-5.303 5.303 1.414 1.414 5.303-5.303 5.303 5.303 1.414-1.414-5.303-5.303z"
                      id="x"
                    />
                  </g>
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
